### ISSUE TITLE
Uninitialized memory performance

### ISSUE THREAD

**Issue by ppetrushkov** [intel/mkl-dnn#672]
I noticed while benchmarking MatMul primitive, that leaving the weight memory uninitialized (as returned by an [empty memory object](https://intel.github.io/mkl-dnn/structdnnl_1_1memory.html#aa28e9a81e89a97b90b49535dc3a5c96e) ) leads to faster execution than initialized memory (zero, random or otherwise). It happens at least with int8 and float types (difference is larger for float). This doesn't happen when input memory in uninitialized.

This isn't really a bug, since one should not be using uninitialized memory in the first place, but I thought you might have an idea why this could happen? It's a bit unintuitive that just by setting the values of an otherwise valid memory object one can run into such performance differences.

One hypothesis is that the compiler optimizes away some of the computation, since the result would be undefined behaviour anyway, could that be the case?

If you'd like to reproduce this, I prepared a minimal example (using DNNL v1.2):
```
#include <iostream>
#include <chrono>
#include <cmath>

#include "dnnl.hpp"


template<typename T>
void init_random(int size, T* ptr) {
    for (int i=0; i<size; ++i)
        ptr[i] = static_cast<T>(127.0f*static_cast<float>(rand()) / static_cast <float> (RAND_MAX));
}

template<typename T>
void init_zero(int size, T* ptr) {
    for (int i=0; i<size; ++i)
        ptr[i] = static_cast<T>(0);
}

int main() {
    int m = 10;
    int n = 3072;
    int k = 768;
    int iters = 1000;

    char trans_a = 'N';
    char trans_b = 'N';
    char offsetc = 'F';
    float alpha = 1.0f;
    int lda = k;
    uint8_t ao = 0;
    int ldb = n;
    int8_t bo = 0;
    float beta = 0.0f;
    int ldc = n;
    std::array<int32_t, 1> oc = {0};

    void* weights_p = std::malloc(k*n*sizeof(int8_t));

    int8_t* weights = static_cast<int8_t*>(weights_p);
    //Uncomment to see the difference
    //init_random(k*n, weights);
    //init_zero(k*n, weights);
    void* input_p = std::malloc(m*k*sizeof(uint8_t));
    uint8_t* input = static_cast<uint8_t*>(input_p);
    init_random(m*k, input);
    void* output_p = std::malloc(m*n*sizeof(int));
    int* output = static_cast<int*>(output_p);
    
    float total=0.0f;
    int dryrun=3;
    for (int i=0; i<iters; ++i) {
        std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
        dnnl_gemm_u8s8s32(trans_a, trans_b, offsetc,
                m, n, k, alpha, const_cast<const uint8_t*>(input), lda, ao,
                const_cast<const int8_t*>(weights), ldb, bo,
                beta, output, ldc, oc.data());
        std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
        if (i>=dryrun)
            total += std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count();
    }
    std::cout << "Average time = " << total/(iters-dryrun) << std::endl;
}   
```

**Comment by emfomenk** [comment]
Hi @ppetrushkov,

Do you use single thread all multiple threads?
In the latter case, do you observe the difference if you force the application to use only one thread?

**Comment by ppetrushkov** [comment]
Hi @emfomenk the difference is most clearly visible when using a single thread, but it seems to also happen when using multiple.

**Comment by emfomenk** [comment]
Hi @ppetrushkov,

Thanks for the reproducer and the issue. It was interesting to dig into it.

> One hypothesis is that the compiler optimizes away some of the computation, since the result would be undefined behaviour anyway, could that be the case?

No, this is definitely not a compiler. When I did all experiments I built the program with `-O0` to make sure that the compiler doesn't interfere with the results.

The culprit is actually a very clever operating system -- guess, like me, you used Linux. In short, the reason you got better performance with uninitialized weights is because all the weights of size `3072*768` ~ 2MB were mapped to one physical page. In other words, the `dnnl_gemm()` accessed only 4KB of memory instead of 2MB. This allows keeping the matrices A, B and C in cache and led to better performance.

Two ways to _fix_ the performance _improvement_ you observed with uninitialized memory:
* Initialize the arrays by writing something to them
* Lock virtual address space into RAM using `mlock` or `mlockall` [1]. The purpose of `mlock` and `mlockall` to actually avoid paging the corresponding memory to the swap area, but as a side effect they map a virtual address space to the physical memory avoiding page overlapping.

To confirm the theory (and also just for fun) I modified your program to print the physical addressed of the weights before and after weights initialization. To print the physical memory address I shamelessly copy-pasted the code from Stack Overflow [2].

``` bash
$  g++ -O0 gh672.cpp -ldnnl && taskset -c 0 ./a.out
Uninitialized memory: 351.07 ms.
Physical addresses: 0x19a8d22000 0x19a8d22000 0x19a8d22000 ... (repeated same address 576 times)

Initialized memory: 480.957 ms.
Physical addresses: 0x1821e7000 0x1d44fe000 0x206d5c000 ... (all 576 addresses are different)
```

If you add `mlock()` (see line 57) for the first case the physical addresses would be different too and time will be the same as in the second case. 

The modified code can be found [here](https://gist.github.com/emfomenk/bc1a8752e31defe9999ee36c4d89150d).

Closing the issue.

--- 

[1] `man 2 mlock`
[2] https://stackoverflow.com/a/45128487
[3] https://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_chapter/libc_3.html
